### PR TITLE
Create an ab initio parse method to quickly make new objects from an MCNP string. 

### LINKED ISSUE #88: Improve ab initio DataCard __init__
Users would like a stop gap measure to add special DataCards like `nps 2e9` without needing full support for the `nps` card yet.

Right now to do this the user would have to make a string for the card put in a list, generate a new `Card`, and then use that `Card` to generate a new `DataCard`.

### PR DESCRIPTION
**Pull request by MicahGale**
# Description

This feature adds the ability to parse MCNP inputs arbitrarily. All `MCNP_Object` now accepts and parses `str`. This also introduces `MCNP_Problem.parse` which is a convenience function that just parses, and goes. The `append` option controls if this new object is added to the problem or not. This also adds the ability to pass a number to a constructor to simplify the build.

Fixes #88 

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

----
📚 Documentation preview 📚: https://montepy--595.org.readthedocs.build/en/595/

### CODE REVIEW

**Comment by MicahGale** [comment]
Also an open question is: if this should ship for 0.6 or 1.0

**Comment by MicahGale** [comment]
To document pretty_str or not?

File: doc/source/changelog.rst [hunk 1]
```diff
@@ -22,6 +21,8 @@ MontePy Changelog
 * Made ``Material.is_atom_fraction`` settable (:issue:`511`). 
 * Made NumberedObjectCollections act like a set (:issue:`138`).
 * Automatically added children objects, e.g., the surfaces in a cell, to the problem when the cell is added to the problem (:issue:`63`).
+* Added ability to parse all MCNP objects from a string (:pull:`595`).
+* Added function: :func:`~montepy.mcnp_problem.MCNP_Problem.parse` to parse arbitrary MCNP object (:pull:`595`).
```
Comment (tjlaboss) [inline doc/source/changelog.rst:25]: Is there a reason these link to a `:pull:` and not an `:issue:`?

File: montepy/mcnp_object.py [hunk 2]
```diff
@@ -100,22 +101,34 @@ class MCNP_Object(ABC, metaclass=_ExceptionContextAdder):
...
-    :type input: Input
+    :type input: Union[Input, str]
     :param parser: The parser object to parse the input with.
-    :type parser: MCNP_Lexer
+    :type parser: MCNP_Parser
+    """
+
+    """
+    The block type this input comes from.
     """
```
Comment (tjlaboss) [inline montepy/mcnp_object.py:111]: Is this separate from the docstring intentionally?

File: montepy/data_inputs/data_input.py [hunk 3]
```diff
@@ -52,17 +54,29 @@ class DataInputAbstract(MCNP_Object):
 
     _classifier_parser = ClassifierParser()
 
-    def __init__(self, input=None, fast_parse=False):
+    def __init__(
+        self,
+        input: Union[montepy.input_parser.mcnp_input.Input, str] = None,
```
Comment (tjlaboss) [inline montepy/data_inputs/data_input.py:59]: Throughout:

The `Union[montepy.input_parser.mcnp_input.Input, str]` is reused frequently.

Consider creating and reusing this type somewhere. For example:

```python
"""
montepy typing
"""
import typing
from montepy.input_parser.mcnp_input import Input

T_InitInput = typing.Union[Input, str]
```

File: montepy/mcnp_object.py [hunk 4]
```diff
@@ -100,22 +101,34 @@ class MCNP_Object(ABC, metaclass=_ExceptionContextAdder):
...
+            self._BLOCK_TYPE
+        except AttributeError:
+            self._BLOCK_TYPE = montepy.input_parser.block_type.BlockType.DATA
         self._problem_ref = None
         self._parameters = ParametersNode()
         self._input = None
         if input:
-            if not isinstance(input, montepy.input_parser.mcnp_input.Input):
+            if not isinstance(input, (montepy.input_parser.mcnp_input.Input, str)):
                 raise TypeError("input must be an Input")
```
Comment (tjlaboss) [inline montepy/mcnp_object.py:127]: " or str"

**Review by tjlaboss — COMMENTED** [review]
I want to monkey around with this some on some complex models and see if I can break it.

Here are some comments in the meanwhile.

File: doc/source/changelog.rst (hunk 1 above)
Comment (MicahGale) [reply doc/source/changelog.rst:25]: Yeah not a good reason. I thought maybe the PR might give more info, but I think it's not a ton of info here. I'll change it.

File: montepy/mcnp_object.py (hunk 2 above)
Comment (MicahGale) [reply montepy/mcnp_object.py:111]: Oh I need to clean that up. It was a doc string for a class variable I had to move.

File: montepy/mcnp_object.py (hunk 4 above)
Comment (MicahGale) [reply montepy/mcnp_object.py:127]: Sometimes I silently ignore optional types in typeErrors to discourage certain behavior, but I think this behavior should be encouraged.

**Comment by MicahGale** [comment]
Overall from a UX perspective, do you like it @tjlaboss?

**Comment by tjlaboss** [comment]
Yes, at first glance, but I want to experiment with it on a practical case.

**Comment by MicahGale** [comment]
Also keep in mind this is merging into alpha-test and not dev. So trying it out can be part of alpha testing.

**Comment by MicahGale** [comment]
I deleted the tests from #597. I think the testing from `mat_redesign` already catches these edge cases.

File: montepy/mcnp_object.py [hunk 5]
```diff
@@ -100,22 +103,30 @@ class MCNP_Object(ABC, metaclass=_ExceptionContextAdder):
...
     """
 
     def __init__(
         self,
-        input: montepy.input_parser.mcnp_input.Input,
+        input: InitInput,
         parser: montepy.input_parser.parser_base.MCNP_Parser,
     ):
+        try:
+            self._BLOCK_TYPE
```
Comment (tjlaboss) [inline montepy/mcnp_object.py:117]: This is `__init__()`: will this branch ever occur?

Comment (MicahGale) [reply montepy/mcnp_object.py:117]: Yes. Subclasses. This is weird to provide a default value that subclasses can override. It has be set inside of init to avoid circular imports.

Comment (tjlaboss) [reply montepy/mcnp_object.py:117]: weird 👍🏻

Comment (MicahGale) [reply montepy/mcnp_object.py:117]: You mean awesome implementation.

**Review by tjlaboss — CHANGES_REQUESTED** [review]
A very nice and useful feature.

See the comments below for one thing to fix, some necessary documentation, and one question.

File: montepy/surfaces/surface.py [hunk 6]
```diff
@@ -1,32 +1,48 @@
...
+    :type number: int
     """
 
     _parser = SurfaceParser()
 
-    def __init__(self, input=None):
-        super().__init__(input, self._parser)
+    def __init__(
+        self,
+        input: InitInput = None,
```
[the remark this answers is not in the document]
Comment (MicahGale) [reply montepy/surfaces/surface.py:36]: Check `surface.old_transform_number` how can you expect a dissociated object to find the transform in a problem?

File: montepy/mcnp_problem.py [hunk 7]
```diff
@@ -623,3 +623,55 @@ def __repr__(self):
                 ret += f"{obj}\n"
             ret += "\n"
         return ret
+
+    def parse(self, input: str):
+        """
+        Parses the MCNP object given by the string, and links it adds it to this problem.
```
[the remark this answers is not in the document]
Comment (MicahGale) [reply montepy/mcnp_problem.py:629]: Hmm interesting. I don't see why you would want the problem side without using `__init__` but don't want append though?

File: montepy/surfaces/surface.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tjlaboss) [reply montepy/surfaces/surface.py:36]: I see. This isn't intuitive IMO. Perhaps the `__repr__` should hint at that. 

Since the information isn't getting lost when instantiating a `Surface` that way, I'll resolve this conversation and open a feature request for revealing more information about dissociated objects.

File: montepy/mcnp_problem.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tjlaboss) [reply montepy/mcnp_problem.py:629]: Basically, it creates an object linked to the problem and with pointers updated, possessing information that instantiating `montepy.Surface`/ `montepy.Cell` doesn't. This would let the user create a base cell from a string and then modify it Pythonically, or do hacky things like writing it in a specific place.

On the other hand, you can just `.pop()` it in one more line of code, so this isn't necessary.

File: montepy/surfaces/surface.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tjlaboss) [reply montepy/surfaces/surface.py:36]: #632

File: montepy/mcnp_problem.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MicahGale) [reply montepy/mcnp_problem.py:629]: I think `pop` is too many steps. Let's add `append`.
